### ISSUE TITLE
Drush 9 cannot SSH into Drupal VM

### ISSUE THREAD

**Issue by danepowell** [geerlingguy/drupal-vm#1595]
### Issue Type

 - Bug Report / Support Request

### Your OS

 - **Linux** (Ubuntu 17.10)

### Summary

DrupalVM provides a template Drush alias file so that Drush knows how to access the VM via SSH:
https://github.com/geerlingguy/drupal-vm/blob/master/provisioning/templates/drupalvm.aliases.drushrc.php.j2

There are a few problems with this for Drush 9:
- drushrc files are replaced with YAML files
- the ssh-options key becomes an array key:
```
ssh:
  options:
```
- because it's a yaml file and not a php file, you can't run commands inline like drush_server_home()

BLT has started converting its own alias template for Drush 9 but I'm stuck on that last step. I figure we should decide on a collective approach to this.
https://github.com/acquia/blt/blob/9.1.x/scripts/drupal-vm/drupal-vm.aliases.yml

**Comment by geerlingguy** [comment]
Hmm... at a quick glance, I'm not sure if there's a simple backwards-compatible way to make this work well. We might have to either drop the feature (managing an alias file on the host), or make it more complicated and have two different templates for `drush_version` < 9 and `drush_version` >= 9.

**Comment by danepowell** [comment]
This is how I'm going to fix it for BLT. Might be marginally less portable, but I think it will still work in most environments: https://github.com/acquia/blt/pull/2241

**Comment by geerlingguy** [comment]
Where are we with this now? Is the simple fix that we would want to use `${env.home}` and that should fix things? Sorry, lost context for this issue :(

**Comment by geerlingguy** [comment]
I'm also just now learning about [drush-launcher](https://github.com/drush-ops/drush-launcher), and it looks like maybe the old way I globally installed drush globally on my Mac (`brew install drush`) is quite deprecated now.

**Comment by geerlingguy** [comment]
So, I'm now seeing new Drush-related issues in Drupal VM's automated builds, e.g.:

```
TASK [geerlingguy.drupal : Install Drupal with drush.] *************************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": ["/var/www/drupalvm/drupal/vendor/drush/drush/drush", "site-install", "standard", "-y", "--root=/var/www/drupalvm/drupal/web", "--site-name=Drupal", "--account-name=admin", "--account-pass=admin", "--db-url=pgsql://drupal:drupal@localhost/drupal"], "delta": "0:00:00.408622", "end": "2018-01-26 04:27:38.036190", "msg": "non-zero return code", "rc": 1, "start": "2018-01-26 04:27:37.627568", "stderr": " [error]  Failed to create database: Password for user drupal:psql: fe_sendauth: no password supplied ", "stderr_lines": [" [error]  Failed to create database: Password for user drupal:psql: fe_sendauth: no password supplied "], "stdout": "\n // You are about to DROP all tables in your 'drupal' database. Do you want to  \n // continue?: yes.                                                             ", "stdout_lines": ["", " // You are about to DROP all tables in your 'drupal' database. Do you want to  ", " // continue?: yes.  
```

From: https://travis-ci.org/geerlingguy/drupal-vm/jobs/333562827

```
  [Symfony\Component\Console\Exception\CommandNotFoundException]  
  Command "@drupalvm.drupalvm.test" is not defined.               
                                                                  
Drush install fail
```

From: https://travis-ci.org/geerlingguy/drupal-vm/jobs/333562825

**Comment by geerlingguy** [comment]
Looks like the first problem is actually https://github.com/drush-ops/drush/issues/3330

**Comment by geerlingguy** [comment]
Also related: https://github.com/geerlingguy/ansible-role-drush/issues/41

**Comment by heddn** [comment]
The change for `drush_server_home()` seems fixable via https://github.com/drush-ops/drush/pull/3115/files

**Comment by geerlingguy** [comment]
Also see: https://github.com/geerlingguy/drupal-vm/issues/1672 (preliminary work before I will be able to work any further on this ticket).

**Comment by geerlingguy** [comment]
Here's some testing I've done on the host side (with Drush 9.0.0 and launcher inside the VM, as per #1672 WIP):

  1. `brew uninstall --force drush`
  2. `brew install drush-launcher`
  3. `drush status` outside of docroot gives error
  4. `drush status` inside docroot shows site info, but not connection info.
  5. `drush cr` inside docroot gives `[error]  SQLSTATE[HY000] [2002] No such file or directory`
  6. `drush site:alias-convert` inside docroot gives `Error: Call to undefined function drush_server_home() in include() (line 14 of /Users/jeff.geerling/.drush/drupalvm.aliases.drushrc.php).`
  7. After removing the `drush_server_home()` bit, I could convert the aliases, which were stashed into my project in a `drush/` directory.
  8. `drush @drupalvm.test` now tries connecting to the right server, but the SSH options aren't being picked up (get `Permission denied (publickey,password).`).

**Comment by geerlingguy** [comment]
Also, here's the generated alias file:

```yaml
 14:09:10 ~/Sites/drupalvm/drupal/drush/sites $ cat drupalvm.site.yml 
test:
  host: drupalvm.test
  options: {  }
  paths:
    drush-script: /var/www/drupalvm/drupal/vendor/drush/drush/drush
  root: /var/www/drupalvm/drupal/web
  uri: drupalvm.test
  user: vagrant
  ssh:
    options: '-o "SendEnv PHP_IDE_CONFIG PHP_OPTIONS XDEBUG_CONFIG" -o PasswordAuthentication=no -i "${env.home}/.vagrant.d/insecure_private_key"'
```

**Comment by geerlingguy** [comment]
Things are getting... complicated.

Basically, Drush 9 no longer _automatically_ looks inside your user home `~/.drush` folder for aliases... instead, according to [these docs](https://github.com/drush-ops/drush/blob/master/examples/example.site.yml#L63-L77), you will have to create a Drush config file (or update one if it exists) to point it at any global (non-project-specific) directories to scan:

```yaml
drush:
  paths:
    alias-path:
      - '${env.home}/.drush/sites'
```

I'm thinking I may write those contents to the file if a config file doesn't already exist. But if it _does_ then we'll have to ask the user to add the proper alias-path configuration. Annoying... but I guess it can work.

**Comment by geerlingguy** [comment]
Next fun problem: https://github.com/drush-ops/drush/issues/3336

**Comment by geerlingguy** [comment]
For the problem linked above, the issue was having a `.` character in the alias name. Drush < 9 was fine with it, but 9+ can't have special characters (well, at least a `.`). So for a short-term fix, I'm grabbing the first part of the domain (e.g. `host.split('.')[0]`), so if you have:

```
drupalvm.test
```

The alias will be `drupalvm`, meaning when you run a command using the alias you'd need to use `drush @drupalvm.drupalvm [command]`.

I've also updated the documentation to account for this change as part of #1672. This will definitely need to be a minor release... and I'm contemplating making it a major since the Drush integration is essentially different in every way :O

**Comment by geerlingguy** [comment]
This is fixed in `master`, will be in 4.8.0, to be released shortly.

**Comment by joelpittet** [comment]
Totally thought Greg and I fixed this the `drush use` seems to not pickup in drush 9.  I followed the stuff above and `drush @cs8.test status` is working but  `drush use @cs8.test`  I'm wild guessing but think maybe something with the laucher?
